sycl/include/sycl/handler.hpp

```diff
@@ -2576,28 +2600,30 @@ class __SYCL_EXPORT handler {
     // TODO add check:T must be an integral scalar value or a SYCL vector type
     static_assert(isValidTargetForExplicitOp(AccessTarget),
                   "Invalid accessor target for the fill method.");
+    // CG::Fill will result in piEnqueuFillBuffer/Image which requires that mem
+    // data is contiguous. Thus we check range and offset when dim > 1
+    // Images don't allow ranged accessors and are fine.
     if constexpr (isBackendSupportedFillSize(sizeof(T)) &&
-                  (Dims <= 1 || isImageOrImageArray(AccessTarget))) {
-      setType(detail::CG::Fill);
-
-      detail::AccessorBaseHost *AccBase = (detail::AccessorBaseHost *)&Dst;
-      detail::AccessorImplPtr AccImpl = detail::getSyclObjImpl(*AccBase);
-
-      MDstPtr = static_cast<void *>(AccImpl.get());
-      CGData.MRequirements.push_back(AccImpl.get());
-      CGData.MAccStorage.push_back(std::move(AccImpl));
-
-      MPattern.resize(sizeof(T));
-      auto PatternPtr = reinterpret_cast<T *>(MPattern.data());
-      *PatternPtr = Pattern;
+                  ((Dims <= 1) || isImageOrImageArray(AccessTarget))) {
+      StageFillCG(Dst, Pattern);
     } else if constexpr (Dims == 0) {
       // Special case for zero-dim accessors.
       parallel_for<__fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
           range<1>(1), [=](id<1>) { Dst = Pattern; });
     } else {
-      range<Dims> Range = Dst.get_range();
-      parallel_for<__fill<T, Dims, AccessMode, AccessTarget, IsPlaceholder>>(
-          Range, [=](id<Dims> Index) { Dst[Index] = Pattern; });
+      // Dim > 1
+      bool OffsetUsable = (Dst.get_offset() == sycl::id<Dims>{});
```

Comment:
Are we sure that sycl::id<Dims>{} is default-initialized to zero?

Reply:
Yes.  According to the spec, that constructor constructs a SYCL id with the value 0 for each dimension. Table 115, Section 4.9.1.3

Reply:
Thanks!